### PR TITLE
Fix: Indent rule (fixes #1797, fixes #1799, fixes #2248, fixes #2343,…

### LINKED ISSUE #1797: Indent rule: unable to enforce indentation of switch cases [$15]
The following code results in having no errors:

`eslint --reset --no-eslintrc --rule "{indent: 2}" test.js`
```js
switch (a) {
case '1':
b();
break;
default:
c();
break;
}
```

If I enable the `indentSwitchCase` option I get the following errors:
`eslint --reset --no-eslintrc --rule "{indent: [2, 4, {indentSwitchCase: true}]}" test.js`

```sh
test.js
  3:4  error  Expected indentation of 4 characters  indent
  6:4  error  Expected indentation of 4 characters  indent
```

I would expect the following errors:
```sh
test.js
  2:4  error  Expected indentation of 4 characters  indent
  3:8  error  Expected indentation of 8 characters  indent
  4:8  error  Expected indentation of 8 characters  indent
  5:4  error  Expected indentation of 4 characters  indent
  6:8  error  Expected indentation of 8 characters  indent
  7:8  error  Expected indentation of 8 characters  indent
```

It seems like the `indentSwitchCase` option enables the indentation check only of the body of a SwitchCase statement. The Indentation of the whole SwitchCase statement is recognized by the rule itself and only validated to be consistet within an single switch statement. There is no option to confiugre a consistent identation for SwitchCase statements of a whole file.

The other strange behavior with this option is, when I explecitly set the value to `false` it doesn’t forbid the identation for a SwitchCase body, it just ignores those lines.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8393484-indent-rule-unable-to-enforce-indentation-of-switch-cases?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #1799: Indent rule ignores single statement if, while, for, do-while
The following examples results in having no error:

```js
if (a)
c();
else if (b)
d();
else
e();
```

```js
while (a)
b();
```

```js
for (;;)
a();
```

```js
for (a in b)
c();
```

```js
do
a();
while (b);
```

I would expect to have an error if a body statement is not indented.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8393685-indent-rule-ignores-single-statement-if-while-for-do-while?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #1800: Indent rule doesn’t validate variable declarations
The following example has no error:

```js
var a = 1,
b = 2,
c = 3;
```

I would expect the following errors:
```sh
2:0  error  Expected indentation of 4 characters  indent
3:0  error  Expected indentation of 4 characters  indent
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8393826-indent-rule-doesn-t-validate-variable-declarations?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #2248: Indentation issue with switch and if inside a case
eslint v 0.18 ? (demo)

```js
/* eslint semi: 0, indent: [2,2] */
var test = "lel"
var ok
var notOk
switch(test) {
case "test":
  break;

case "assert":
  if(test) {
    ok++
  }
  else {
    notOk++
  }
}
```

```
13:0 - Expected indentation of 0 characters.
14:2 - Expected indentation of 2 characters.
15:0 - Expected indentation of 0 characters.
```

Expected:

No errors.

---

I also tested this snippet with indented cases which give same sort of errors

```js
/* eslint semi: 0, indent: [2,2] */
var test = "lel"
var ok
var notOk
switch(test) {
  case "test":
    break;

  case "assert":
    if(test) {
      ok++
    }
    else {
      notOk++
    }
    break
}
```

```
13:2 - Expected indentation of 2 characters.
14:4 - Expected indentation of 4 characters.
15:2 - Expected indentation of 2 characters.
```

Expected: 

No errors.

### LINKED ISSUE #2278: Indentation of object literals
Not sure if this is considered a feature request or a bug, but I can't find any combination of style rules that would warn about these code snippets:

```
let myvar = {
          y: 1,
            x: 2
                           };
```
or:
```
func(1, 2,
        {
x: {
  y: 3
          }
               });
```

The indentation of these snippets are way off, but as I said, I can't find any rules to flag this. I guess it's not a simple problem since you might want to align function arguments to start after the function name, something like:
```
func(1,
     2,
     {
       x: {
         y: 3
       }
     }
);
```

But it should be pretty universal that `{` and `}` should be on the same indentation level, and that properties in an object literal should have the same indentation.

I'd be happy to take a crack at it, if someone could give me some starting pointers.

### LINKED ISSUE #2343: Incorrect indent when else statement has no curly braces
**The `indent` error is incorrectly raised when an `if` statement uses curly braces, but its `else` counterpart does not.**

The following example has proper indentation:
```JavaScript
/*eslint "indent": 2*/
/*eslint "curly": 0*/
(function () {
    "use strict";
    switch (this.a) {
        case "a":
            if (this.a === this.b) {
                this.a = "a";
                this.b = "b";
            }
            else
                this.a = "c";  // this line triggers an error
            break;
    }
})();
```
but results in
```
12:8  error  Expected indentation of 8 characters  indent
```

The following example has improper indentation:
```JavaScript
/*eslint "indent": 2*/
/*eslint "curly": 0*/
(function () {
    "use strict";
    switch (this.a) {
        case "a":
            if (this.a === this.b) {
                this.a = "a";
                this.b = "b";
            }
            else
        this.a = "c";  // this line should trigger an error
            break;
    }
})();
```
but does not trigger any error.

Even when curly braces are used, the error is off. This examples has proper indentation:

```JavaScript
/*eslint "indent": 2*/
(function () {
    "use strict";
    switch (this.a) {
        case "a":
            if (this.a === this.b) {
                this.a = "a";
                this.b = "b";
            }
            else {
                this.a = "c";
            }
            break;
    }
})();
```
triggers:
```
10:8   error  Expected indentation of 8 characters   indent
11:12  error  Expected indentation of 12 characters  indent
12:8   error  Expected indentation of 8 characters   indent
```

### PR DESCRIPTION
**Pull request by gyandeeps**
… fixes #2278, fixes #1800)

**Important**
Its an improvement over the old rule. I am not saying this is perfect but its way better.
It solves a lot of our issues. I have not changed the options of the rule so it will not be a breaking change.

### CODE REVIEW

**Comment by gyandeeps** [comment]
@puzrin We have inherited your rule and modified and enhanced it. Do let me know if you have any feedback.

**Comment by puzrin** [comment]
No problem. I already said, code can be used in eslint in any way if it helps, and even without mentioning my authorship. Thank you very much for improving upstream.

**Comment by puzrin** [comment]
https://github.com/nodeca/eslint-plugin-nodeca/blob/master/test/fixtures.yml

Make sure all those tests exist or have replacements.

**Comment by gyandeeps** [comment]
I have replacements for almost all but I will double check. Thanks.

File: lib/rules/indent.js [hunk 1]
```diff
@@ -1,53 +1,53 @@
...
 
-    var lines = null;
-    var indentStack = [0];
-    var linesToCheck = null;
-    var breakIndents = null;
-
     if (context.options.length) {
         if (context.options[0] === "tab") {
-            indentChar = "\t";
+            // indentChar = "\t";
```
Comment (nzakas) [inline lib/rules/indent.js]: Safe to remove this?

File: lib/rules/indent.js [hunk 2]
```diff
@@ -56,408 +56,331 @@ module.exports = function (context) {
...
 
     /**
-     * Mark line to be checked
-     * @param {Number} line - line number
-     * @returns {void}
+     * Get node indent
+     * @param {ASTNode} node Node to examine
+     * @param {boolean} byLastLine get indent of node's last line
+     * @param {boolean} excludeCommas skip comma on start of line
+     * @returns {number} Indent
```
Comment (nzakas) [inline lib/rules/indent.js]: `{int}`

**Comment by nzakas** [comment]
Looks awesome!

**Comment by gyandeeps** [comment]
@nzakas @ilyavolodin I think I am all set now.
Do let me know when you guys ready to merge, at that time I will rebase just to make sure nothing is breaking inside master because we use `indent` rule inside our project linting.

**Comment by ilyavolodin** [comment]
@gyandeeps Can you run perf test before and after changes for comparison?

**Comment by gyandeeps** [comment]
**Before**:

```sh
Gyandeep@GYANDEEP ~/WebstormProjects/eslint (master)
$ npm run perf

> eslint@0.24.0 perf C:\Users\Gyandeep\WebstormProjects\eslint
> node Makefile.js perf

CPU Speed is 2394 with multiplier 7500000
Performance Run #1:  2370.239191ms
Performance Run #2:  2716.265021ms
Performance Run #3:  2544.124735ms
Performance Run #4:  2618.084562ms
Performance Run #5:  2435.953415ms
Performance budget ok:  2544.124735ms (limit: 3132.8320802005014ms)
```

**After**:

```sh
Gyandeep@GYANDEEP ~/WebstormProjects/eslint (IndentRuleFix)
$ npm run perf

> eslint@0.24.0 perf C:\Users\Gyandeep\WebstormProjects\eslint
> node Makefile.js perf

CPU Speed is 2394 with multiplier 7500000
Performance Run #1:  2381.150365ms
Performance Run #2:  2344.4879849999998ms
Performance Run #3:  2381.615653ms
Performance Run #4:  2351.355686ms
Performance Run #5:  2354.670434ms
Performance budget ok:  2354.670434ms (limit: 3132.8320802005014ms)
```

**Comment by ilyavolodin** [comment]
Very nice! That's pretty significant boost in performance!

**Comment by nzakas** [comment]
Lgtm. Rebase away!

**Comment by gyandeeps** [comment]
Get Set Go

**Comment by gyandeeps** [comment]
@nzakas Get Set Merge :smile:

**Comment by nzakas** [comment]
And the crowd applauds

**Comment by gyandeeps** [comment]
:smile:  :tada:

**Comment by rhutchison** [comment]
@gyandeeps thanks for your work on this.  What are your thoughts on this case?

```javascript
foo(new Bar({
    debug: true,
    hostname: localhost
  },
  function() {
    //some code
    return;
  }
));
```

**Comment by gyandeeps** [comment]
@rhutchison Can you please log an issue sharing what you see right now with `eslint` and what do you expect?
Thanks

**Pull request merged**